src/spi_comms.c

```diff
@@ -0,0 +1,357 @@
+#include "spi_comms.h"
+
+// SPI data from PAY-SSM
+volatile uint8_t spi_first_byte = 0;
+volatile uint8_t spi_second_byte = 0;
+
+volatile uint8_t spi_spdr_tx_buffer = 0;    // holds data to be shifted into SPDR, for SPI transfer
+volatile uint8_t spi_frame_number = 0;      // byte number of data trasmission
+
+volatile uint8_t spi_status_byte = 0;
+
+
+// Initialize SPI comms as SPI slave, with interrupts enabled
+void init_spi_comms(void){
+    // initialize SPI
+    init_spi();
+
+    // set DATA_RDYn pin as output high
+    DATA_RDYn_DDR |= _BV(DATA_RDYn);    // set direction = output
+    opt_set_data_rdy_high();
+}
+
+
+// return status of SS pin
+uint8_t opt_check_ss_pin(void){
+    return SS_PIN & _BV(SS); 
+}
+
+
+// set DATA_RDYn low
+void opt_set_data_rdy_low(){
+    DATA_RDYn_PORT &= ~_BV(DATA_RDYn);     // set output LOW
+}
+
+// set DATA_RDYn high
+void opt_set_data_rdy_high(){
+    DATA_RDYn_PORT |= _BV(DATA_RDYn);  // set output HIGH
+}
+
+
+// to be put in infinite loop in main
+void opt_loop(void){
+    // if SPI transfer if completed
+    if (SPSR & _BV(SPIF)){
+        spi_first_byte = SPDR;
+        opt_set_data_rdy_low();         // set DATA_RDYn low to signal to PAY that it's ready for the second byte
+
+        // wait until another SPI transfer is completed
+        // --> aka wait until SPIF is no longer high 
+        while (!(SPSR & _BV(SPIF)));
+
+        opt_set_data_rdy_high();        // clear DATA_RDYn, so PAY doesn't freak out
+        spi_second_byte = SPDR;
+
+        // now, got both bytes
+        // perform the requested command and send back data if necessary
+        manage_cmd(spi_first_byte);
+    }
+    opt_set_data_rdy_high();
+}
+
+
+// depending on cmd_code, does appropriate requested function + return data (if needed)
+void manage_cmd (uint8_t cmd_code){
+    // if first byte is get_reading, then 2nd byte is well data
+    if (cmd_code == CMD_GET_READING){
+        // spi_second_byte contains well_data
+        opt_update_reading(spi_second_byte);    // performs reading (3 bytes), stores it in wells[32] of well_t
+        opt_transfer_reading();       // shifts reading data into SPDR over 3 SPI transmissions
+    }
+
+    // invalid command
+    else{ 
+        spi_status_byte |= _BV(SPI_ERROR_BIT) | _BV(SPI_INVALID_COMMAND_BIT);
+        SPDR = spi_status_byte;
+        opt_set_data_rdy_low();
+
+        // wait until SPI transfer is complete
+        while (!(SPSR & _BV(SPIF)));
+        opt_set_data_rdy_high();
+    }
+
+}
+
+
+// calibrate and take well readings
+// well_data[7] - optical density = 0, fluorescent LED = 1
+// well_data[5:0] - well number (0-31)
+void opt_update_reading(uint8_t well_data){
+    update_well_reading((well_data && 0x1F), well_data >> 7);

```

Comment:
Should be `&`, not  `&&`. Also to get the LSB 6 bits, you need `0x3F` instead of `0x1F`.